File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        target = 0
        triplets = set()
        nums = sorted(nums)
        seen = []
        for i in range(len(nums)):
            target = nums[i]
            l, r = i + 1, len(nums) - 1
            if target in seen:
                continue
            elif target > 0:
                break
            else:
                seen.append(target)
                while l < r:
                    if nums[l] + nums[r] == -target:
                        triplet = tuple(sorted((nums[i], nums[l], nums[r])))
                        triplets.add(triplet)
                    if nums[i] + nums[l] + nums[r] < 0:
                        l += 1
                    else:
                        r -= 1
        
        return list(list(x) for x in triplets)

File: 15-3sum/3sum.cpp
#include <ranges> 

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int i = 0, j = 1, k = 2;
        int num_items = nums.size();
        sort(nums.begin(), nums.end());
        


        std::map<int, int> seen_nums;
        for (int i = 0; i < num_items; i++) {
            seen_nums[nums[i]] = i;
        }

        std::vector<std::vector<int>> result;
        for (int i = 0; i < num_items; i++) {
            if (nums[i] > 0) {
                break;
            }
            for (int j = i + 1; j < num_items; j++) {
                int required = -1 * (nums[i] + nums[j]);
                if (seen_nums.contains(required) && seen_nums[required] > j) {
                    result.push_back({nums[i], nums[j], required});
                }
                j = seen_nums[nums[j]];
            }
            i = seen_nums[nums[i]];
        }
    

        return result;
    }
};